tools/pythonpkg/pyduckdb/filesystem.py

```diff
@@ -0,0 +1,91 @@
+from fsspec import filesystem, AbstractFileSystem
+from fsspec.implementations.memory import MemoryFileSystem
+from shutil import copyfileobj
+from io import StringIO, TextIOBase
+
+# Shamelessly stolen from pandas
+class BytesIOWrapper:
+	# Wrapper that wraps a StringIO buffer and reads bytes from it
+	# Created for compat with pyarrow read_csv
+	def __init__(self, buffer: StringIO | TextIOBase, encoding: str = "utf-8") -> None:
+		self.buffer = buffer
+		self.encoding = encoding
+		# Because a character can be represented by more than 1 byte,
+		# it is possible that reading will produce more bytes than n
+		# We store the extra bytes in this overflow variable, and append the
+		# overflow to the front of the bytestring the next time reading is performed
+		self.overflow = b""
+
+	def __getattr__(self, attr: str):
+		return getattr(self.buffer, attr)
+
+	def read(self, n: int | None = -1) -> bytes:
+		assert self.buffer is not None
+		bytestring = self.buffer.read(n).encode(self.encoding)
+		#When n=-1/n greater than remaining bytes: Read entire file/rest of file
+		combined_bytestring = self.overflow + bytestring
+		if n is None or n < 0 or n >= len(combined_bytestring):
+			self.overflow = b""
+			return combined_bytestring
+		else:
+			to_return = combined_bytestring[:n]
+			self.overflow = combined_bytestring[n:]
+			return to_return
+
+def is_file_like(obj):
```

Comment:
Can we not use `isinstance(obj, IOBase)` for this? Or is that too strict?

Reply:
Stolen from pandas, they do this, so yea I think it's too strict?

Reply:
Fair enough yeah - do we need to attribute back to them, given we stole code?

Reply:
Not entirely sure what you mean by that, but sure?
We can probably make it less strict, I don't think we need `write` or `__iter__` for our usecase

Reply:
But their BytesIOWrapper is a nice solution

Reply:
Yea I'm moved their BytesIOWrapper solution to its own file with the license attached 👍